p5rpc.modloader/Merging/SpdMerger.cs

```diff
@@ -0,0 +1,245 @@
+        if (_mergedFileCache.TryGet(mergedKey, innerSources, out var mergedCachePath))
+        {
+            _logger.Info("Loading Merged SPD {0} from Cache ({1})", route, mergedCachePath);
+            _utils.ReplaceFileInBinderInput(pathToFileMap, route, mergedCachePath);
+            return;
+        }
+
+        // Try and get cached original file
+        var originalKey = MergedFileCache.CreateKey(pathInCpk, modIds);
+
+        if (!_mergedFileCache.TryGet(originalKey, cpkSources, out var cachedPath))
+        {
+            // Else Merge our Data
+            // First we extract.
+            await Task.Run(async () =>
+            {
+                await using var cpkStream = new FileStream(cpkPath, FileMode.Open, FileAccess.ReadWrite, FileShare.ReadWrite);
+                using var reader = _criFsApi.GetCriFsLib().CreateCpkReader(cpkStream, false);
+                using var extractedFile = reader.ExtractFile(cpkEntry.Files[fileIndex].File);
+                var item = await _mergedFileCache.AddAsync(originalKey, cpkSources, extractedFile.RawArray.AsMemory(0, extractedFile.Count));
+                cachedPath = Path.Combine(_mergedFileCache.CacheFolder, item.RelativePath);
+            });
+        }
+
+        string spdPath = Path.Combine(bindDirectory, route);
+        string? dir = Path.GetDirectoryName(spdPath);
+        if (dir != null)
+            Directory.CreateDirectory(dir);
+
+        if (!_spdEmulator.TryCreateFromFileSlice(cachedPath!, 0, pathInCpk, spdPath))
+        {
+            _logger.Error($"Cannot Create File From Slice!");
+            return;
+        }
+
+        // Cache merged
+        var item = await _mergedFileCache.AddAsync(mergedKey, innerSources, await File.ReadAllBytesAsync(spdPath));
+        _utils.ReplaceFileInBinderInput(pathToFileMap, route, spdPath);
+        _logger.Info("Merge {0} Complete. Cached to {1}.", route, item.RelativePath);
+    }
+
+    private async ValueTask CachePakedSpd(string route, string[] cpks, CachedFileSource[] cpkSources, PakSpdRoutes innerFiles)
+    {
+        // Try and get cached merged spd
+        string[] modIds = { "p5rpc.modloader" };
+        var mergedKey = MergedFileCache.CreateKey(route, modIds);
+
+        CachedFileSource[] spdSources = innerFiles.SpdRoutes.Select(file => new CachedFileSource { LastWrite = File.GetLastWriteTime(file) }).ToArray();
+
+        DateTime lastWrite = DateTime.MinValue;
+        foreach (var source in spdSources)
+            if (source.LastWrite > lastWrite) lastWrite = source.LastWrite;
+
+        if (_mergedFileCache.TryGet(mergedKey, spdSources, out var mergedCachePath))
+        {
+            _logger.Info("Loading Merged SPD {0} from Cache ({1})", route, mergedCachePath);
+            foreach (var path in innerFiles.PakRoutes)
+            {
+                _spdEmulator.RegisterSpd(mergedCachePath, path);
+                File.SetLastWriteTime(path, lastWrite);
```

Comment:
The `RegisterVirtualFile` method which is used under the hood by `RegisterSpd` always creates a new file. Therefore, the last write time should (hopefully) already be set there.